[Android] Add HDR allowed dynamic metadata formats setting

quietvoid opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
The changes add a new Expert setting in the Player/Video settings, named `Allowed HDR dynamic metadata formats`.
The setting allows users to choose the preferred dynamic HDR format for playback, or none.

It works by completely removing the metadata from the bitstream prior to decoding.
For Dolby Vision, the NAL types can just be filtered out.
For HDR10+, it is necessary to parse the SEI messages and remove the SMPTE ST 2094-40 / HDR10+ one.
Supports HEVC only.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Improves compatibility problems with videos containing both Dolby Vision and HDR10+ on some devices (FireTV Stick).
As dynamic HDR videos are becoming more and more common, a lot of people encounter issues playing them because the hardware has major bugs that cannot be worked around through Android APIs.

Furthermore, webOS on some TV models also needs this setting: https://github.com/xbmc/xbmc/issues/24390
It will be added once this is accepted.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
There's a custom build on the forums: https://forum.kodi.tv/showthread.php?tid=371557

The Dolby Vision removing code has been in use for a while.
The HDR10+ code has been reworked but the SEI parsing code should be OK and mostly matches other implementations like in `FFmpeg`.

Tested in runtime on FireTV Stick 4K Max gen 1 and 2.
With a display supporting both Dolby Vision/HDR10+, the display switches modes correctly and the image looks fine.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
No effect by default.
Users must manually select the HDR dynamic metadata to keep, otherwise the videos are left untouched.

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

quietvoid committed:
[Android] Allow removing Dolby Vision dynamic HDR metadata

quietvoid committed:
Utils: Add HEVC SEI parser

quietvoid committed:
[Android] Allow removing HDR10+ dynamic HDR metadata

thexai commented on xbmc/utils/HevcSei.cpp:
```diff
@@ -0,0 +1,186 @@
+    if (i > 2 && buf[i - 2] == 0 && buf[i - 1] == 0 && buf[i] <= 3)
+    {
+      buf.insert(buf.begin() + i, 3);
+      count += 1;
```
`count` variable seems redundant as is increased at same time as `buf.insert`

Use directly `while (i < buf.size())` should be equivalent.

quietvoid replied:
Yes it is the same. I wasn't sure if `buf.size()` is as performant.

thexai commented on xbmc/utils/HevcSei.cpp:
```diff
@@ -0,0 +1,186 @@
+    if (sei.payload_type == 4 && sei.payload_size >= 7)
+    {
+      CBitstreamReader br(buf.data() + sei.payload_offset, sei.payload_size);
+      auto itu_t_t35_country_code = br.ReadBits(8);
```
```suggestion
      const auto itu_t_t35_country_code = br.ReadBits(8);
```

This and others

thexai commented on xbmc/utils/HevcSei.h:
```diff
@@ -0,0 +1,58 @@
+  CHevcSei() = default;
+  ~CHevcSei() = default;
+
+  uint8_t payload_type;
```
```suggestion
  uint8_t payload_type{0};
```

This and others. 
Initializing inside the class seems safer in order to avoid "missing initialization" mistakes when class is used.

thexai reviewed (commented):
Code seems in a very good shape. Only a minor suggestions.

quietvoid commented:
Addressed comments.
As well as:
- Moved `CHevcSei::ParseSeiMessage` to private visibility.
- Removed `last_payload_{type,size}_byte` from `CHevcSei` member variables, since they're only needed for parsing.

neo1973 commented on xbmc/utils/HevcSei.h:
```diff
@@ -0,0 +1,54 @@
+  size_t payload_offset{0};
+
+  // Parses SEI payload assumed to not have emulation prevention 3 bytes
+  static const std::vector<CHevcSei> ParseSeiRbsp(const uint8_t* buf, const size_t len);
```
Please remove the `const` here and the following functions when returning by value, see http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-return-const

quietvoid replied:
Thanks. Done.

sundermann commented on xbmc/utils/HevcSei.cpp:
```diff
@@ -0,0 +1,187 @@
+
+#include "HevcSei.h"
+
+void hevc_add_start_code_emulation_prevention_3_byte(std::vector<uint8_t>& buf)
```
Pascal case here and in other places

sundermann commented on xbmc/utils/HevcSei.cpp:
```diff
@@ -0,0 +1,187 @@
+    return 1;
+
+  br.SkipBits(sei.payload_size * 8);
+  messages.push_back(sei);
```
```suggestion
  messages.emplace_back(sei);
```

sundermann commented on xbmc/utils/HevcSei.cpp:
```diff
@@ -0,0 +1,187 @@
+  {
+    out.reserve(len);
+
+    out.push_back(buf[0]);
```
```suggestion
    out.emplace_back(buf[0]);
```

sundermann commented on xbmc/utils/HevcSei.cpp:
```diff
@@ -0,0 +1,187 @@
+    out.reserve(len);
+
+    out.push_back(buf[0]);
+    out.push_back(buf[1]);
```
```suggestion
    out.emplace_back(buf[1]);
```

sundermann commented on xbmc/utils/HevcSei.cpp:
```diff
@@ -0,0 +1,187 @@
+    for (i = 2; i < len; i++)
+    {
+      if (!(buf[i - 2] == 0 && buf[i - 1] == 0 && buf[i] == 3))
+        out.push_back(buf[i]);
```
```suggestion
        out.emplace_back(buf[i]);
```

sundermann commented on xbmc/utils/HevcSei.cpp:
```diff
@@ -0,0 +1,187 @@
+    auto msg = *res;
+
+    containsHdr10Plus = true;
+    if (messages.size() > 1)
```
```suggestion
    if (!messages.isEmpty())
```

quietvoid replied:
Not the same, it looks for > 1.

sundermann replied:
Alright!

thexai replied:
Do you mean `messages.empty()` ?

Anyway is not equivalent (`!empty()` is same as `> 0` but code condition is > 1)

sundermann reviewed (changes requested):
The variable/parameter/method naming is inconsistent. Please follow the coding guidelines for newly added code even if there's other code in the same file following a different naming convention.

thexai commented on xbmc/utils/HevcSei.cpp:
```diff
@@ -0,0 +1,187 @@
+
+int CHevcSei::ParseSeiMessage(CBitstreamReader& br, std::vector<CHevcSei>& messages)
+{
+  CHevcSei sei{};
```
```suggestion
  CHevcSei sei;
```

As now class is self-initialized is not necessary here

quietvoid commented on xbmc/utils/BitstreamConverter.cpp:
```diff
@@ -356,6 +357,8 @@ CBitstreamConverter::CBitstreamConverter()
   m_sps_pps_context.sps_pps_data = NULL;
   m_start_decode = true;
   m_convert_dovi = false;
+  m_remove_dovi = false;
```
What about those like `itu_t_t35_country_code` that follow spec syntax?
That one also doesn't camel case very well.

sundermann replied:
No strong opinion on those, maybe someone else from the team knows what to do in those cases

fritsch replied:
In terms of "matches given class" I would keep m_remove_dovi as it is.

fritsch replied:
Edit: Seen it was changed below already ... also fine with me. There are two difficult things for computer scientis:
- Naming things
- Cache Invalidation
- Off by one errors

quietvoid commented:
Fixed:
- Variables casing and member variables prefixing with `m_`.
- Function naming to PascalCase.
- Replaced use of `push_back` with `emplace_back`.

garbear commented on xbmc/utils/HevcSei.h:
```diff
@@ -0,0 +1,54 @@
+#include <optional>
+#include <vector>
+
+class CHevcSei
```
A new reader might come across this class and not know what HEVC or SEI is. GPT-4 suggests some helpful documentation:

```c++
/*!
 * \brief Parses HEVC SEI messages for supplemental video information.
 *
 * The CHevcSei class is used to interpret and handle Supplemental Enhancement
 * Information (SEI) messages found in High Efficiency Video Coding (HEVC)
 * bitstreams. It is particularly useful for extracting HDR10+ metadata and
 * other types of supplemental data from HEVC encoded video streams.
 *
 * \note This class deals with SEI messages in HEVC streams and does not
 * process the video content itself.
 */
 ```

fritsch replied:
Yeah, please add that, so we have it. Not sure someone will read the class (Bitstreamconverter) and asking himself about that, but it's good to have it.

When that comment is in - I am - for my part - perfectly fine with the PR. Thanks very much! This is a missing addition for v21, as it actually is a bugfix, cause without it  - some people cannot watch HDR10+ or DV content due to a black screen. I am one of them.

quietvoid replied:
I agree. Added it.

fritsch reviewed (approved):
Please add garbear's Documentation suggestion, then it is good too go from my side.

fritsch commented:
@sundermann the PascalCase okay with you?

thexai reviewed (approved):
Runtime tested on Shield Pro:

NOTE: Shield filters by itself HDR10+ so not need this to able playback hybrid DolbyVision / HDR10+ files anyway is interesting to test compatibility and possible regressions.

Tested with a file DV + HDR10+:
`Dolby Vision, Version 1.0, dvhe.08.06, BL+RPU, HDR10 compatible / SMPTE ST 2094 App 4, Version 1, HDR10+ Profile B compatible`

1. DV enabled, HDR10+ enabled: as expected plays same as before (stream untouched).
2. DV enabled, HDR10+ disabled: plays same as before (Shield not uses HDR10+ metadata).
3. DV disabled, HDR10+ enabled: plays as regular HDR10 (Shield not supports HDR10+).
4. DV disabled, HDR10+ disabled: plays as regular HDR10.

Conclusion: it works perfectly fine on Shield and this test shows that stream is processed correctly.

fritsch commented:
@garbear @neo1973 All your comments adressed properly?

garbear reviewed (approved):
Not that familiar with HDR but code looks fine

fritsch commented:
@neo1973  just push it in when you are happy.

Thanks much @quietvoid for yet another immense contribution!

kjvalencik commented:
@quietvoid thanks for providing this workaround! I'm fairly new to the kodi code base.

I'm trying to understand, is there a reason these settings were made to *always* filter instead of only when both exist? Would it be feasible to "sniff" the file and provide preferences?

In other words:
* Only HDR+ present. Use HDR+.
* Only DV present. Use DV.
* Both HDR+ & DV present. Filter HDR+, use DV.

Thanks!

quietvoid commented:
Not easily mainly because they are per frame metadata.

kjvalencik commented:
I'm not sure if this is a good way to go about it, but inserting `removeHdr10Plus = true;` [here](https://github.com/xbmc/xbmc/blob/348313a4335d14be88c8c91e70bdb3a67a5d542b/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp#L538) served my use case of removing HDR10+ if both are present, but keeping it if there's no dolby vision.

quietvoid commented:
Yes but that only works if Dolby Vision is detected at the container level.

It can't be done the same way to disable Dolby Vision when HDR10+ is present.

KarellenX commented:
Hello @quietvoid 

I can't see this setting on my system. Would you mind providing a 1920x1080 screenshot of the settings page so I can add to the wiki here... https://kodi.wiki/view/Settings/Player/Videos

Thanks.

KarellenX commented:
Thank you @quietvoid

matthuisman commented:
any simple way to set this programatically from an add-on using a listitem property or something?

I have some users of addons where the HDR isnt "good" and they want to disable HDR just for this addon.

I guess I could set the kodi setting but then I need to make sure to revert it (hacky)

quietvoid commented:
This does not allow to disable HDR, so it's not what youre looking for.

matthuisman commented:
These are single hevc streams with HDR. But work on sdr too (backwards
compatible). When the user has hdr compatible hardware, we still want the
sdr.

What does this PR do?

On Thu, 5 Dec 2024, 08:01 quietvoid, ***@***.***> wrote:

> HDR support is mostly the OS, you ideally you need the addon to not
> request an HDR stream.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/24584#issuecomment-2518319804>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABPQAKOHQO6Y7B3DJK3VWB32D5GQRAVCNFSM6AAAAABS7UXWBOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMJYGMYTSOBQGQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

quietvoid commented:
> These are single hevc streams with HDR. But work on sdr too (backwards
> compatible).

I don't understand what you mean, HLG?
HDR10 streams are not backwards compatible.

> What does this PR do?

Only strips SEI from the bitstream.
